src/main/java/com/squareup/kotlinpoet/TypeSpec.kt

```diff
@@ -441,6 +444,14 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
     fun addSuperinterface(superinterface: KClass<*>)
         = addSuperinterface(superinterface.asTypeName())
 
+    fun addSuperinterface(superinterface: ParameterSpec) = apply {
```

Comment:
`addSuperinterface(superinterface: KClass<*>, delegateParameterName: String)` is much more straightforward, but there will need to be assertions made about what the parameter will be generated as (and consequently - restrictions)